### PR TITLE
Optimize delegate allocations in SocketTaskExtensions

### LINKED ISSUE #4052: System.Net.Sockets TPL tracking items
- [ ] Optimize delegate allocations. 
- [x] Investigate behavior difference from Desktop: Used to throw SocketException.
- [x] Investigate behavior difference in tests in System.Net.Sockets.Legacy

### PR DESCRIPTION
**Pull request by Maxwe11**
Fixes #4052 

If you have plans to eliminate Begin/End in short terms feel free to close PR :smile:

### CODE REVIEW

**Comment by Maxwe11** [comment]
If you have plans to eliminate Begin/End in short terms feel free to close PR :smile:

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketTaskExtensions.cs [hunk 1]
```diff
@@ -66,69 +101,108 @@ public static Task<int> ReceiveAsync(this Socket socket, ArraySegment<byte> buff
...
+                    var arguments = (Tuple<Socket, EndPoint>)state;
+                    Socket s = arguments.Item1;
+                    EndPoint e = arguments.Item2;
+
+                    return s.BeginReceiveFrom(
+                        targetBuffer.Array,
+                        targetBuffer.Offset,
+                        targetBuffer.Count,
+                        flags,
+                        ref e,
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketTaskExtensions.cs]: This doesn't look correct.  This ref is changing the local e, which is going to be a copy of what's in the tuple, so the end delegate is going to potentially using the wrong e, right?

Same in the later similar transformations.

If I'm right, this also suggests there's a test gap.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketTaskExtensions.cs [hunk 2]
```diff
@@ -66,69 +101,108 @@ public static Task<int> ReceiveAsync(this Socket socket, ArraySegment<byte> buff
...
+                    Socket s = arguments.Item1;
+                    SocketFlags f = arguments.Item2;
+                    EndPoint e = arguments.Item3;
+
+                    return s.BeginReceiveMessageFrom(
+                        targetBuffer.Array,
+                        targetBuffer.Offset,
+                        targetBuffer.Count,
+                        f,
+                        ref e,
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketTaskExtensions.cs]: See previous comments.

**Comment by stephentoub** [comment]
A few concerns, but the rest looks good.

**Comment by stephentoub** [comment]
cc: @davidsh, @CIPop

**Comment by davidsh** [comment]
cc: @josguil

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketTaskExtensions.cs (hunk 1 above)
Comment (Maxwe11) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketTaskExtensions.cs]: Yes, you are right.
About test gap, [seems it's intentionally](https://github.com/dotnet/corefx/blob/3669a883eae0ee87f2db20f88005a26b825f0dc2/src/System.Net.Sockets.Legacy/tests/FunctionalTests/DualModeSocketTest.cs#L1768L1784).
`BeginReceiveFrom` in tests it's a helper extensions which fallback to `ReceiveFromAsync` extension.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketTaskExtensions.cs (hunk 2 above)
Comment (Maxwe11) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketTaskExtensions.cs]: Same as above.

**Comment by josguil** [comment]
@Maxwe11  Thanks for addressing this TODO item on #4052 .

**Comment by davidsh** [comment]
LGTM

**Pull request merged**